src/late_multicellular_stage/DualContourer.cs

```diff
@@ -0,0 +1,423 @@
+                });
+            });
+        });
+
+        sw.Stop();
+    }
+
+    private void SetColours(List<Vector3> points, Color[] colours)
+    {
+        for (int i = 0; i < colours.Length; i++)
+        {
+            colours[i] = MathFunction!.GetColour(points[i]);
+        }
+    }
+
+    private int IdOf(bool x0Y0Z0, bool x0Y0Z1, bool x0Y1Z0, bool x0Y1Z1, bool x1Y0Z0, bool x1Y0Z1, bool x1Y1Z0,
+        bool x1Y1Z1)
+    {
+        int id = 0;
+
+        if (x0Y0Z0)
+            id += 1;
+        if (x0Y0Z1)
+            id += 2;
+        if (x0Y1Z0)
+            id += 4;
+        if (x0Y1Z1)
+            id += 8;
+        if (x1Y0Z0)
+            id += 16;
+        if (x1Y0Z1)
+            id += 32;
+        if (x1Y1Z0)
+            id += 64;
+        if (x1Y1Z1)
+            id += 128;
+
+        return id;
+    }
+
+    private Vector3I[]? GetBestTriangles(int x, int y, int z, bool[,,] shapePoints)
+    {
+        int id = IdOf(shapePoints[x, y, z],
+            shapePoints[x, y, z + 1],
+            shapePoints[x, y + 1, z],
+            shapePoints[x, y + 1, z + 1],
+            shapePoints[x + 1, y, z],
+            shapePoints[x + 1, y, z + 1],
+            shapePoints[x + 1, y + 1, z],
+            shapePoints[x + 1, y + 1, z + 1]);
+
+        lookupTableInt!.TryGetValue(id, out var result);
+
+        return result;
+    }
+
+    private Vector3 GetFunctionMomentarySpeed(Vector3 realPos, float funcAtPoint, float d = 0.01f)
+    {
+        Vector3 direction = new Vector3(
+            (MathFunction!.GetValue(new Vector3(realPos.X + d, realPos.Y, realPos.Z)) - funcAtPoint) / d,
+            (MathFunction.GetValue(new Vector3(realPos.X, realPos.Y + d, realPos.Z)) - funcAtPoint) / d,
+            (MathFunction.GetValue(new Vector3(realPos.X, realPos.Y, realPos.Z + d)) - funcAtPoint) / d);
+
+        return direction;
+    }
+
+    /// <summary>
+    ///   Adjusts mesh points to be at mesh's surface. Optionally also calculates mesh normals.
+    /// </summary>
+    private void AdjustVertices(List<Vector3> points, float changeClamp = 0.5f, Vector3[]? meshNormals = null)
+    {
+        // Vertex adjustment by normals:
+        // 1. Find functions instanteous speed at vertex
+        // 2. Find required distance change for vertex to be at the mesh's surface
+        // 3. Clamp the change for it not to be too far from the initial position.
+        // In other words, no vertex should end up in a place where another vertex theoretically may be.
+        // 4. Apply change, using instanteous speed's vector as direction
+        // 5. Make a second pass of 1-4. This algorithm is only an approximation, so sometimes one-pass approach
+        // results in spiky meshes. Optionally clamp the value more than on the first pass.
+
+        float maxToleratedChange = changeClamp / PointsPerUnit;
+
+        float d = 0.25f / PointsPerUnit;
+
+        int count = points.Count;
+        Parallel.For(0, count, delegate(int i)
```

Comment:
This should use our TaskExecutor as we don't use the default threading system. (also applies to anywhere else in case you added threading elsewhere as well in this PR).

Reply:
Done